util/decimal: compute decimal.Log with a Taylor series

madelynnblue opened the pull request:
Using the Taylor series reduces test time by half, I think because
it more quickly converges due to the input range reduction to 0.9 <
x < 1.1.

maxPrecision has been reduced here which, by itself, also reduced
test times by half because large computations more quickly detected
and aborted.

Overall test time for this package went from 25s to 6s.

testFuncTimeout could go as low as 400ms or so, but it is kept at 1s
to prevent flaky tests.


name           old time/op  new time/op  delta
DecimalLog-4    157µs ± 2%   143µs ± 7%   -8.71%  (p=0.008 n=5+5)
DecimalPow-4  1.31ms ± 8%  1.09ms ± 7%  -16.52%  (p=0.036 n=5+3)

See ln_var() in: https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numeric.c

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/11669)
<!-- Reviewable:end -->

arjunravinarayan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
I did not hand-check your math, and would prefer if there were more tests, but that bit is not your introduction, so lgtm. In general the only thing missing are a few explanatory comments for variables you introduce.

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/util/decimal/decimal.go, line 290 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXkkPzjF0O49miDmcLi:-KXkkPzjF0O49miDmcLj:bggj4b8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L290)):*
> ```Go
> 	}
> 
> 	ns := s * 2
> ```

I think we now need a comment as to what the purpose of `ns` is, now that all the surrounding context is removed. Also, why is `2` the right value? It's not clear to me that it is.

---

*[pkg/util/decimal/decimal.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXkkryZ2OFuOL4-ZM41:-KXkkryZ2OFuOL4-ZM42:bubfq9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L299)):*
> ```Go
> 	}
> 
> 	fact := inf.NewDec(2, 0)
> ```

Add comment stating that `fact` keeps track of how how much we've mutated the input to get it within the desired range.

---

*[pkg/util/decimal/decimal.go, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXklUHLT7Fa6lj2Xu04:-KXklUHLT7Fa6lj2Xu05:b-bbyj3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L307)):*
> ```Go
> 
> 	// Reduce input range of x to 0.9 < x < 1.1 by taking the square root to
> 	// improve convergence rate.
> ```

change to something like 
//The taylor series of ln(x) converges much faster if x until 0.9 < x < 1.1. Thus, successively square-root x until it is in that region. Keep track of how many square-rootings were done using fact.

---

*[pkg/util/decimal/decimal.go, line 330 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXkngmCojo59PRWMUZI:-KXkngmCojo59PRWMUZJ:bu86xq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L330)):*
> ```Go
> 		numerator.Mul(numerator, elem)
> 		div := inf.NewDec(int64(loop.i)*2+3, 0)
> 		v.QuoRound(numerator, div, ns, inf.RoundHalfUp)
> ```

I'd generally add comment lines (like you did for elem before the start of the loop) before every math statement. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-:-KXkmwIWgONnlNyis0jy:b-sqziug)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> other than the comments brought up by Arjun and the questions about reducing decimal allocations.

---

Reviewed 2 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/util/decimal/decimal.go, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXklUHLT7Fa6lj2Xu04:-KXl3_iccyLAlMy2PH0y:b2f20ps) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L307)):*
> if x until

Is this missing a word or is the "until" extraneous?

---

*[pkg/util/decimal/decimal.go, line 315 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXl3wxbv07XsB1MEk19:-KXl3wxbv07XsB1MEk1A:b-qeky2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L315)):*
> ```Go
> 	xp1 := new(inf.Dec)
> 	xp1.Add(x, decimalOne)
> 	xm1 := new(inf.Dec)
> ```

Are all of these allocations needed? Seems like some of these `inf.Dec` objects can be reused.

---

*[pkg/util/decimal/decimal.go, line 329 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXl4QJfq1tb3NKo6X03:-KXl4QJfq1tb3NKo6X04:bspwka5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L329)):*
> ```Go
> 	for loop := newLoop("log", x, s, 40); ; {
> 		numerator.Mul(numerator, elem)
> 		div := inf.NewDec(int64(loop.i)*2+3, 0)
> ```

There's no way to avoid the allocation in the loop?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-:-KXl4bDuUnLcRrt9j0an:b-q2eg09)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
util/decimal: compute decimal.Log with a Taylor series

Using the Taylor series reduces test time by half, I think because
it more quickly converges due to the input range reduction to 0.9 <
x < 1.1.

maxPrecision has been reduced here which, by itself, also reduced
test times by half because large computations more quickly detected
and aborted.

Overall test time for this package went from 25s to 6s.

testFuncTimeout could go as low as 400ms or so, but it is kept at 1s
to prevent flaky tests.


name           old time/op  new time/op  delta
DecimalLog-4    157µs ± 2%   143µs ± 7%   -8.71%  (p=0.008 n=5+5)
DecimalPow-4  1.31ms ± 8%  1.09ms ± 7%  -16.52%  (p=0.036 n=5+3)

See ln_var() in: https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numeric.c

mjibson commented:
Review status: 1 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/util/decimal/decimal.go, line 290 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXkkPzjF0O49miDmcLi:-KXqhGKSj4dI45KgHOvh:b-3v8zk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L290)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I think we now need a comment as to what the purpose of `ns` is, now that all the surrounding context is removed. Also, why is `2` the right value? It's not clear to me that it is.
</blockquote></details>

I've simplified how this is done and added a comment describing the current accuracy wrt postgres. This will be improved later (that is, we will understand better how this works) with the upcoming new decimal package.

---

*[pkg/util/decimal/decimal.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXkkryZ2OFuOL4-ZM41:-KXlmovyuMflbKF9_M12:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L299)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Add comment stating that `fact` keeps track of how how much we've mutated the input to get it within the desired range.
</blockquote></details>

Done.

---

*[pkg/util/decimal/decimal.go, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXklUHLT7Fa6lj2Xu04:-KXlmnXhexcwdl8zoVGL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L307)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> if x until

Is this missing a word or is the "until" extraneous?
</blockquote></details>

Done.

---

*[pkg/util/decimal/decimal.go, line 315 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXl3wxbv07XsB1MEk19:-KXlmoAkaQbIRp04r6lV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L315)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are all of these allocations needed? Seems like some of these `inf.Dec` objects can be reused.
</blockquote></details>

Done.

---

*[pkg/util/decimal/decimal.go, line 329 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXl4QJfq1tb3NKo6X03:-KXlmmu8ZzQ0wt8obe2h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L329)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

There's no way to avoid the allocation in the loop?
</blockquote></details>

Done.

---

*[pkg/util/decimal/decimal.go, line 330 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669#-KXkngmCojo59PRWMUZI:-KXlmmj61KMbXqsNGM9G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2dd0460c2bda4e3b5797f9efd747fffae55a241/pkg/util/decimal/decimal.go#L330)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I'd generally add comment lines (like you did for elem before the start of the loop) before every math statement. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11669)*
<!-- Sent from Reviewable.io -->
